functions/api/action.js

```diff
@@ -0,0 +1,208 @@
+}
+
+module.exports = router;
+
+/**
+ * Method to create a new transaction
+ * @param {data for the transaction} data
+ */
+module.exports.createTransaction = async function (data) {
+    const branchId = data[constants.BRANCH]
+    delete data[constants.BRANCH]
+    const branchRef = db.collection(constants.BRANCHES).doc(branchId);
+    const branchSnapshot = await branchRef.get()
+    if (!branchSnapshot.exists) {
+        const error = new Error(`Requested branch is not present. Transaction will not go through.`)
+        error.statusCode = 404
+        next(error)
+        return;
+    }
+
+    // Get the last transaction for the product to get the intialQuantity for current transaction
+    const transactionRef = branchRef.collection(constants.TRANSACTIONS)
+    try {
+        const lastTransaction = await transactionRef
+            .where(constants.PRODUCT, '==', data[constants.PRODUCT])
+            .orderBy(constants.DATE, 'desc')
+            .limit(1)
+            .get();
+
+        let initialQuantity = 0
+        lastTransaction.forEach((transaction) => {
+            initialQuantity = transaction.data()[constants.CLOSING_QUANTITY]
+        })
+        data[constants.INITIAL_QUANTITY] = initialQuantity
+        data[constants.CLOSING_QUANTITY] = getClosingQuantity(data[constants.OPERATION], initialQuantity,
+            data[constants.OPERATIONAL_QUANTITY])
+        const resultTransaction = await transactionRef.add(data)
+        return resultTransaction.id
+    } catch (err) {
+        logger.error(err)
+        throw err
+    }
+}
+
+/**
+ * Method to calculate closing quantity for the current transaction
+ * @param {} operation
+ * @param {*} initialQuantity
+ * @param {*} operationalQuantity
+ */
+function getClosingQuantity(operation, initialQuantity, operationalQuantity) {
+    switch (operation) {
+        case constants.ADD_PRODUCT:
+        case constants.TRANSFER_IN:
+            return initialQuantity + operationalQuantity
+        case constants.ISSUE_PRODUCT:
+        case constants.TRANSFER_OUT:
+            return initialQuantity - operationalQuantity
+        case constants.ADJUSTMENT:
+            return operationalQuantity
+    }
+}
+
+/**
+ * Trigger to update availableQuantity and isBelowThreshold value after
+ * a transaction is completed for a particular branch
+ */
+module.exports.updateAvailableQuantityInInventory = functions.firestore
+    .document(`/${constants.BRANCHES}/{branch}/${constants.TRANSACTIONS}/{transactionId}`)
+    .onWrite(async (snapshot, context) => {
+        const transactionRecord = snapshot.after.data()
+        const branchId = context.params.branch
+        const productId = transactionRecord[constants.PRODUCT]
+        logger.info(`Updating availableQuantity for product ${transactionRecord[constants.PRODUCT]} in branch ${branchId}`)
+
+        const inventoryRef = await db.collection(constants.BRANCHES).doc(branchId)
+                                .collection(constants.INVENTORY)
```

Comment:
dint you make inventory as top level collection?

Reply:
No, each branch has a inventory. I made it as a subcollection